sd/etcd/client_test.go

```diff
@@ -62,3 +65,249 @@ func TestOptions(t *testing.T) {
 		t.Errorf("expected error: %v", err)
 	}
 }
+
+// Mocks of the underlying etcd.KeysAPI interface that is called by the methods we want to test
+
+// fakeKeysAPI implements etcd.KeysAPI, event and err are channels used to emulate
+// an etcd event or error, getres will be returned when etcd.KeysAPI.Get is called.
+type fakeKeysAPI struct {
+	event  chan bool
+	err    chan bool
+	getres *getResult
+}
+
+type getResult struct {
+	resp *etcd.Response
+	err  error
+}
+
+// Get return the content of getres or nil, nil
+func (fka *fakeKeysAPI) Get(ctx context.Context, key string, opts *etcd.GetOptions) (*etcd.Response, error) {
+	if fka.getres == nil {
+		return nil, nil
+	}
+	return fka.getres.resp, fka.getres.err
+}
+
+// Set is not used in the tests
+func (fka *fakeKeysAPI) Set(ctx context.Context, key, value string, opts *etcd.SetOptions) (*etcd.Response, error) {
+	return nil, nil
+}
+
+// Delete is not used in the tests
+func (fka *fakeKeysAPI) Delete(ctx context.Context, key string, opts *etcd.DeleteOptions) (*etcd.Response, error) {
+	return nil, nil
+}
+
+// Create is not used in the tests
+func (fka *fakeKeysAPI) Create(ctx context.Context, key, value string) (*etcd.Response, error) {
+	return nil, nil
+}
+
+// CreateInOrder is not used in the tests
+func (fka *fakeKeysAPI) CreateInOrder(ctx context.Context, dir, value string, opts *etcd.CreateInOrderOptions) (*etcd.Response, error) {
+	return nil, nil
+}
+
+// Update is not used in the tests
+func (fka *fakeKeysAPI) Update(ctx context.Context, key, value string) (*etcd.Response, error) {
+	return nil, nil
+}
+
+// Watcher return a fakeWatcher that will forward event and error received on the channels
+func (fka *fakeKeysAPI) Watcher(key string, opts *etcd.WatcherOptions) etcd.Watcher {
+	return &fakeWatcher{fka.event, fka.err}
+}
+
+// fakeWatcher implements etcd.Watcher
+type fakeWatcher struct {
+	event chan bool
+	err   chan bool
+}
+
+// Next blocks until an etcd event or error is emulated.
+// When an event occurs it just return nil response and error.
+// When an error occur it return a non nil error.
+func (fw *fakeWatcher) Next(context.Context) (*etcd.Response, error) {
+	for {
+		select {
+		case <-fw.event:
+			return nil, nil
+		case <-fw.err:
+			return nil, errors.New("error from underlying etcd watcher")
+		default:
+		}
+	}
+}
+
+// getfakeClient return a new etcd.Client built on top of the mocked interfaces
+func getFakeClient(event, err chan bool, getres *getResult) (Client, error) {
+	return &client{
+		keysAPI: &fakeKeysAPI{event, err, getres},
+		ctx:     context.Background(),
+	}, nil
+}
```

Comment:
No reason to have the return signature include an error that's always nil, unless I'm missing something?

```go
func newFakeClient(...) Client {
   return &client{...}
}
```